File: nbody.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

// Conditional Include for OpenMP
#ifdef _OPENMP
#include <omp.h>
#else
#include <time.h>
#include <sys/time.h>
#endif

typedef struct {
    float x, y, z;
    float vx, vy, vz;
    float mass;
} Body;

// Cross-platform timer function
double get_time() {
    #ifdef _OPENMP
        return omp_get_wtime();
    #else
        struct timeval tv;
        gettimeofday(&tv, NULL);
        return tv.tv_sec + tv.tv_usec * 1e-6;
    #endif
}

void update_physics(Body* bodies, int N, float dt, int steps) {
    for (int t = 0; t < steps; t++) {
        // Only parallelize if OpenMP is active
        #pragma omp parallel for schedule(static)
        for (int i = 0; i < N; i++) {
            float Fx = 0, Fy = 0, Fz = 0;
            for (int j = 0; j < N; j++) {
                if (i == j) continue;
                float dx = bodies[j].x - bodies[i].x;
                float dy = bodies[j].y - bodies[i].y;
                float dz = bodies[j].z - bodies[i].z;
                float dist_sq = dx*dx + dy*dy + dz*dz + 1e-9f;
                float dist = sqrtf(dist_sq);
                float f = (1.0f * bodies[i].mass * bodies[j].mass) / (dist_sq * dist);
                Fx += f * dx;
                Fy += f * dy;
                Fz += f * dz;
            }
            bodies[i].vx += Fx * dt / bodies[i].mass;
            bodies[i].vy += Fy * dt / bodies[i].mass;
            bodies[i].vz += Fz * dt / bodies[i].mass;
        }

        for (int i = 0; i < N; i++) {
            bodies[i].x += bodies[i].vx * dt;
            bodies[i].y += bodies[i].vy * dt;
            bodies[i].z += bodies[i].vz * dt;
        }
    }
}

int main(int argc, char** argv) {
    if (argc != 3) { printf("Usage: %s <N> <steps>\n", argv[0]); return 1; }
    int N = atoi(argv[1]);
    int steps = atoi(argv[2]);
    float dt = 0.01f;

    Body* bodies = (Body*)malloc(N * sizeof(Body));
    
    for (int i = 0; i < N; i++) {
        bodies[i].x = (float)(rand() % 100);
        bodies[i].y = (float)(rand() % 100);
        bodies[i].z = (float)(rand() % 100);
        bodies[i].vx = 0; bodies[i].vy = 0; bodies[i].vz = 0;
        bodies[i].mass = 1.0f;
    }

    double start = get_time();
    update_physics(bodies, N, dt, steps);
    double end = get_time();

    printf("%f", end - start);
    free(bodies);
    return 0;
}

File: nbody.py
import numpy as np
import time
import sys
from numba import njit, prange

# --- 1. Pure Python (Very Slow, for baseline) ---
def update_physics_pure(pos, vel, mass, dt, steps):
    N = len(pos)
    for _ in range(steps):
        for i in range(N):
            Fx, Fy, Fz = 0.0, 0.0, 0.0
            for j in range(N):
                if i == j: continue
                dx = pos[j,0] - pos[i,0]
                dy = pos[j,1] - pos[i,1]
                dz = pos[j,2] - pos[i,2]
                dist_sq = dx*dx + dy*dy + dz*dz + 1e-9
                dist = dist_sq ** 0.5
                f = (mass[i] * mass[j]) / (dist_sq * dist)
                Fx += f * dx
                Fy += f * dy
                Fz += f * dz
            vel[i,0] += Fx * dt / mass[i]
            vel[i,1] += Fy * dt / mass[i]
            vel[i,2] += Fz * dt / mass[i]
        pos += vel * dt

# --- 2. Numba Parallel (Uses OpenMP backend) ---
@njit(parallel=True, fastmath=True)
def update_physics_numba(pos, vel, mass, dt, steps):
    N = len(pos)
    for _ in range(steps):
        # prange explicitly tells Numba to parallelize this loop (like #pragma omp parallel for)
        for i in prange(N):
            Fx, Fy, Fz = 0.0, 0.0, 0.0
            for j in range(N):
                if i == j: continue
                dx = pos[j,0] - pos[i,0]
                dy = pos[j,1] - pos[i,1]
                dz = pos[j,2] - pos[i,2]
                dist_sq = dx*dx + dy*dy + dz*dz + 1e-9
                dist = dist_sq ** 0.5
                f = (mass[i] * mass[j]) / (dist_sq * dist)
                Fx += f * dx
                Fy += f * dy
                Fz += f * dz
            vel[i,0] += Fx * dt / mass[i]
            vel[i,1] += Fy * dt / mass[i]
            vel[i,2] += Fz * dt / mass[i]
        
        # Parallelize position update too
        for i in prange(N):
            pos[i,0] += vel[i,0] * dt
            pos[i,1] += vel[i,1] * dt
            pos[i,2] += vel[i,2] * dt

if __name__ == "__main__":
    mode = sys.argv[1] # "pure" or "numba"
    N = int(sys.argv[2])
    steps = int(sys.argv[3])
    
    pos = np.random.rand(N, 3).astype(np.float32)
    vel = np.zeros((N, 3), dtype=np.float32)
    mass = np.ones(N, dtype=np.float32)
    
    start = time.time()
    if mode == "pure":
        update_physics_pure(pos, vel, mass, 0.01, steps)
    elif mode == "numba":
        # First call includes compilation time, so we run a dummy first
        update_physics_numba(pos, vel, mass, 0.01, 1) 
        start = time.time() # Reset timer
        update_physics_numba(pos, vel, mass, 0.01, steps)
        
    print(time.time() - start)